### ISSUE TITLE
cpprepl seems to be completely broken

### ISSUE THREAD

**Issue by 71walceli** [Enter-tainer/cpprepl#1]
Whenever I try to do anything, I get a lot of errors, for example:

```
apk@HP-ProBook-6470b ~ $ cpprepl --g++
✔ >>= · int a=1;
In file included from /usr/include/c++/5/cstdint:35:0,
                 from /tmp/dbg.h:31,
                 from /tmp/repl.cxx:1:
/usr/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
 #error This file requires compiler and library support \
  ^
In file included from /tmp/repl.cxx:1:0:
/tmp/dbg.h:52:8: error: ‘constexpr’ does not name a type
 static constexpr size_t PREFIX_LENGTH =
        ^
/tmp/dbg.h:52:8: note: C++11 ‘constexpr’ only available with -std=c++11 or -std=gnu++11
/tmp/dbg.h:54:8: error: ‘constexpr’ does not name a type
 static constexpr size_t SUFFIX_LENGTH = sizeof("]") - 1;
        ^
/tmp/dbg.h:54:8: note: C++11 ‘constexpr’ only available with -std=c++11 or -std=gnu++11
/tmp/dbg.h: In function ‘std::__cxx11::string mgt::get_type_name(mgt::type_tag<T>)’:
/tmp/dbg.h:73:22: error: ‘PREFIX_LENGTH’ is not a member of ‘pf’
   return type.substr(pf::PREFIX_LENGTH,
                      ^
/tmp/dbg.h:74:36: error: ‘PREFIX_LENGTH’ is not a member of ‘pf’
                      type.size() - pf::PREFIX_LENGTH - pf::SUFFIX_LENGTH);
                                    ^
/tmp/dbg.h:74:56: error: ‘SUFFIX_LENGTH’ is not a member of ‘pf’
                      type.size() - pf::PREFIX_LENGTH - pf::SUFFIX_LENGTH);
                                                        ^
/tmp/dbg.h: In function ‘std::__cxx11::string mgt::type_name()’:
/tmp/dbg.h:78:7: error: ‘is_volatile’ is not a member of ‘std’
   if (std::is_volatile<T>::value) {
       ^
/tmp/dbg.h:78:25: error: expected primary-expression before ‘>’ token
   if (std::is_volatile<T>::value) {
                         ^
/tmp/dbg.h:78:26: error: ‘::value’ has not been declared
   if (std::is_volatile<T>::value) {
                          ^
/tmp/dbg.h:79:9: error: ‘is_pointer’ is not a member of ‘std’
     if (std::is_pointer<T>::value) {
         ^
/tmp/dbg.h:79:26: error: expected primary-expression before ‘>’ token
     if (std::is_pointer<T>::value) {
                          ^
/tmp/dbg.h:79:27: error: ‘::value’ has not been declared
     if (std::is_pointer<T>::value) {
                           ^
/tmp/dbg.h:80:38: error: ‘remove_volatile’ in namespace ‘std’ does not name a template type
       return type_name<typename std::remove_volatile<T>::type>() + " volatile";
                                      ^
/tmp/dbg.h:80:53: error: expected template-argument before ‘<’ token
       return type_name<typename std::remove_volatile<T>::type>() + " volatile";
                                                     ^
/tmp/dbg.h:80:53: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:80:63: error: expected class-name before ‘(’ token
       return type_name<typename std::remove_volatile<T>::type>() + " volatile";
                                                               ^
/tmp/dbg.h:80:63: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:80:63: error: expected identifier before ‘(’ token
/tmp/dbg.h:82:52: error: ‘remove_volatile’ in namespace ‘std’ does not name a template type
       return "volatile " + type_name<typename std::remove_volatile<T>::type>();
                                                    ^
/tmp/dbg.h:82:67: error: expected template-argument before ‘<’ token
       return "volatile " + type_name<typename std::remove_volatile<T>::type>();
                                                                   ^
/tmp/dbg.h:82:67: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:82:77: error: expected class-name before ‘(’ token
       return "volatile " + type_name<typename std::remove_volatile<T>::type>();
                                                                             ^
/tmp/dbg.h:82:77: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:82:77: error: expected identifier before ‘(’ token
/tmp/dbg.h:85:7: error: ‘is_const’ is not a member of ‘std’
   if (std::is_const<T>::value) {
       ^
/tmp/dbg.h:85:22: error: expected primary-expression before ‘>’ token
   if (std::is_const<T>::value) {
                      ^
/tmp/dbg.h:85:23: error: ‘::value’ has not been declared
   if (std::is_const<T>::value) {
                       ^
/tmp/dbg.h:86:9: error: ‘is_pointer’ is not a member of ‘std’
     if (std::is_pointer<T>::value) {
         ^
/tmp/dbg.h:86:26: error: expected primary-expression before ‘>’ token
     if (std::is_pointer<T>::value) {
                          ^
/tmp/dbg.h:86:27: error: ‘::value’ has not been declared
     if (std::is_pointer<T>::value) {
                           ^
/tmp/dbg.h:87:38: error: ‘remove_const’ in namespace ‘std’ does not name a template type
       return type_name<typename std::remove_const<T>::type>() + " const";
                                      ^
/tmp/dbg.h:87:50: error: expected template-argument before ‘<’ token
       return type_name<typename std::remove_const<T>::type>() + " const";
                                                  ^
/tmp/dbg.h:87:50: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:87:60: error: expected class-name before ‘(’ token
       return type_name<typename std::remove_const<T>::type>() + " const";
                                                            ^
/tmp/dbg.h:87:60: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:87:60: error: expected identifier before ‘(’ token
/tmp/dbg.h:89:49: error: ‘remove_const’ in namespace ‘std’ does not name a template type
       return "const " + type_name<typename std::remove_const<T>::type>();
                                                 ^
/tmp/dbg.h:89:61: error: expected template-argument before ‘<’ token
       return "const " + type_name<typename std::remove_const<T>::type>();
                                                             ^
/tmp/dbg.h:89:61: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:89:71: error: expected class-name before ‘(’ token
       return "const " + type_name<typename std::remove_const<T>::type>();
                                                                       ^
/tmp/dbg.h:89:71: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:89:71: error: expected identifier before ‘(’ token
/tmp/dbg.h:92:7: error: ‘is_pointer’ is not a member of ‘std’
   if (std::is_pointer<T>::value) {
       ^
/tmp/dbg.h:92:24: error: expected primary-expression before ‘>’ token
   if (std::is_pointer<T>::value) {
                        ^
/tmp/dbg.h:92:25: error: ‘::value’ has not been declared
   if (std::is_pointer<T>::value) {
                         ^
/tmp/dbg.h:93:36: error: ‘remove_pointer’ in namespace ‘std’ does not name a template type
     return type_name<typename std::remove_pointer<T>::type>() + "*";
                                    ^
/tmp/dbg.h:93:50: error: expected template-argument before ‘<’ token
     return type_name<typename std::remove_pointer<T>::type>() + "*";
                                                  ^
/tmp/dbg.h:93:50: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:93:60: error: expected class-name before ‘(’ token
     return type_name<typename std::remove_pointer<T>::type>() + "*";
                                                            ^
/tmp/dbg.h:93:60: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:93:60: error: expected identifier before ‘(’ token
/tmp/dbg.h:95:7: error: ‘is_lvalue_reference’ is not a member of ‘std’
   if (std::is_lvalue_reference<T>::value) {
       ^
/tmp/dbg.h:95:33: error: expected primary-expression before ‘>’ token
   if (std::is_lvalue_reference<T>::value) {
                                 ^
/tmp/dbg.h:95:34: error: ‘::value’ has not been declared
   if (std::is_lvalue_reference<T>::value) {
                                  ^
/tmp/dbg.h:96:36: error: ‘remove_reference’ in namespace ‘std’ does not name a template type
     return type_name<typename std::remove_reference<T>::type>() + "&";
                                    ^
/tmp/dbg.h:96:52: error: expected template-argument before ‘<’ token
     return type_name<typename std::remove_reference<T>::type>() + "&";
                                                    ^
/tmp/dbg.h:96:52: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:96:62: error: expected class-name before ‘(’ token
     return type_name<typename std::remove_reference<T>::type>() + "&";
                                                              ^
/tmp/dbg.h:96:62: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:96:62: error: expected identifier before ‘(’ token
/tmp/dbg.h:98:7: error: ‘is_rvalue_reference’ is not a member of ‘std’
   if (std::is_rvalue_reference<T>::value) {
       ^
/tmp/dbg.h:98:33: error: expected primary-expression before ‘>’ token
   if (std::is_rvalue_reference<T>::value) {
                                 ^
/tmp/dbg.h:98:34: error: ‘::value’ has not been declared
   if (std::is_rvalue_reference<T>::value) {
                                  ^
/tmp/dbg.h:99:36: error: ‘remove_reference’ in namespace ‘std’ does not name a template type
     return type_name<typename std::remove_reference<T>::type>() + "&&";
                                    ^
/tmp/dbg.h:99:52: error: expected template-argument before ‘<’ token
     return type_name<typename std::remove_reference<T>::type>() + "&&";
                                                    ^
/tmp/dbg.h:99:52: error: expected ‘>’ before ‘<’ token
/tmp/dbg.h:99:62: error: expected class-name before ‘(’ token
     return type_name<typename std::remove_reference<T>::type>() + "&&";
                                                              ^
/tmp/dbg.h:99:62: error: expected ‘::’ before ‘(’ token
/tmp/dbg.h:99:62: error: expected identifier before ‘(’ token
/tmp/dbg.h: In function ‘void mgt::print_bytes(T)’:
/tmp/dbg.h:107:8: error: ‘f’ does not name a type
   auto f        = cout.flags();
        ^
/tmp/dbg.h:108:8: error: ‘p’ does not name a type
   auto p        = reinterpret_cast<uint8_t*>(&val);
        ^
/tmp/dbg.h:110:3: error: ‘uint32_t’ was not declared in this scope
   uint32_t step = sizeof(T);
   ^
/tmp/dbg.h:112:20: error: type/value mismatch at argument 1 in template parameter list for ‘template<class _Tp, class _Alloc> class std::vector’
     vector<uint32_t> res;
                    ^
/tmp/dbg.h:112:20: note:   expected a type, got ‘uint32_t’
/tmp/dbg.h:112:20: error: template argument 2 is invalid
/tmp/dbg.h:113:25: error: ‘step’ was not declared in this scope
     for (int i = 0; i < step; ++i)
                         ^
/tmp/dbg.h:114:11: error: request for member ‘push_back’ in ‘res’, which is of non-class type ‘int’
       res.push_back(p[i]);
           ^
/tmp/dbg.h:114:21: error: ‘p’ was not declared in this scope
       res.push_back(p[i]);
                     ^
/tmp/dbg.h:115:17: error: request for member ‘begin’ in ‘res’, which is of non-class type ‘int’
     reverse(res.begin(), res.end());
                 ^
/tmp/dbg.h:115:30: error: request for member ‘end’ in ‘res’, which is of non-class type ‘int’
     reverse(res.begin(), res.end());
                              ^
/tmp/dbg.h:117:17: error: ‘p’ does not name a type
       for (auto p : res) {
                 ^
/tmp/dbg.h:122:5: error: expected ‘;’ before ‘}’ token
     } else {
     ^
/tmp/dbg.h:122:5: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:122:5: error: expected ‘;’ before ‘}’ token
/tmp/dbg.h:122:5: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:122:5: error: expected ‘)’ before ‘}’ token
/tmp/dbg.h:122:5: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:123:17: error: ‘p’ does not name a type
       for (auto p : res)
                 ^
/tmp/dbg.h:125:5: error: expected ‘;’ before ‘}’ token
     }
     ^
/tmp/dbg.h:125:5: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:125:5: error: expected ‘;’ before ‘}’ token
/tmp/dbg.h:125:5: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:125:5: error: expected ‘)’ before ‘}’ token
/tmp/dbg.h:125:5: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:127:23: error: ‘res’ was not declared in this scope
     vector<bitset<8>> res;
                       ^
/tmp/dbg.h:127:23: error: template argument 1 is invalid
/tmp/dbg.h:127:12: error: template argument 1 is invalid
     vector<bitset<8>> res;
            ^
/tmp/dbg.h:127:12: error: template argument 2 is invalid
/tmp/dbg.h:128:25: error: ‘step’ was not declared in this scope
     for (int i = 0; i < step; ++i)
                         ^
/tmp/dbg.h:129:32: error: ‘uint8_t’ was not declared in this scope
       res.push_back(bitset<8>((uint8_t)p[i]));
                                ^
/tmp/dbg.h:131:15: error: ‘p’ does not name a type
     for (auto p : res)
               ^
/tmp/dbg.h:133:3: error: expected ‘;’ before ‘}’ token
   }
   ^
/tmp/dbg.h:133:3: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:133:3: error: expected ‘;’ before ‘}’ token
/tmp/dbg.h:133:3: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:133:3: error: expected ‘)’ before ‘}’ token
/tmp/dbg.h:133:3: error: expected primary-expression before ‘}’ token
/tmp/dbg.h:135:14: error: ‘f’ was not declared in this scope
   cout.flags(f);
              ^
/tmp/dbg.h: At global scope:
/tmp/dbg.h:138:70: error: expected class-name before ‘{’ token
 template <typename, typename = void> struct is_stl : std::false_type {};
                                                                      ^
/tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
 struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
                  ^
/tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
/tmp/dbg.h:141:39: error: ‘declval’ is not a member of ‘std’
 struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
                                       ^
/tmp/dbg.h:141:57: error: ‘.’ cannot appear in a constant-expression
 struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
                                                         ^
/tmp/dbg.h:141:63: error: a function call cannot appear in a constant-expression
 struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
                                                               ^
/tmp/dbg.h:142:39: error: ‘declval’ is not a member of ‘std’
                                       std::declval<T>().end())>>
                                       ^
/tmp/dbg.h:142:62: error: a function call cannot appear in a constant-expression
                                       std::declval<T>().end())>>
                                                              ^
/tmp/dbg.h:143:12: error: template argument 2 is invalid
     : std::true_type {};
            ^
/tmp/dbg.h:148:15: error: ‘enable_if’ in namespace ‘std’ does not name a template type
 typename std::enable_if<!is_stl<T>::value, std::string>::type
               ^
/tmp/dbg.h:148:24: error: expected unqualified-id before ‘<’ token
 typename std::enable_if<!is_stl<T>::value, std::string>::type
                        ^
/tmp/dbg.h:154:15: error: ‘enable_if’ in namespace ‘std’ does not name a template type
 typename std::enable_if<is_stl<T>::value, std::string>::type
               ^
/tmp/dbg.h:154:24: error: expected unqualified-id before ‘<’ token
 typename std::enable_if<is_stl<T>::value, std::string>::type
                        ^
```

Anything will cause that many errors. 

I've no problem whatsoever using `g++` itself to compile code. Since I don't have `clang` installed, it will say it's not found.

Some useful data
OS: Linux Mint 18.3
g++ verson: g++ (Ubuntu 5.5.0-12ubuntu1~16.04) 5.5.0 20171010

It may not be a bug in recent versions of g++, but apt won't update it, so I can't discard that possibility. Also, no requirements seem to be specified, so I don't know...

Thanks in advance...

**Comment by Enter-tainer** [comment]
Are you using the latest version of this package? If you install it with `npm i cpprepl -g`, you can run `npm ls -g cpprepl` to show its version. The latest version is 1.2.4.

```
> npm ls -g cpprepl
/usr/lib
└── cpprepl@1.2.4
```

**Comment by 71walceli** [comment]
I've got 1.2.3
```
/usr/lib
└── cpprepl@1.2.3 
```

I think I installed at most two weeks ago, so didn't think of it...

I'll try to upgrade it...

**Comment by 71walceli** [comment]
Now I updated, yet I seem to have another problem:

```
✔ >>= · int a=5;
In file included from /tmp/repl.cxx:1:0:
/tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
 struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
                  ^
/tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
/tmp/dbg.h:142:63: error: template argument 2 is invalid
```

**Comment by Enter-tainer** [comment]
Update may fix this bug, since 1.2.4 explicitly add `-std=c++17` option. 

But I guess bugs will still occur because maybe gcc 5 is too old to fully support c++17.

**Comment by Enter-tainer** [comment]
Is is possible for you to upgrade gcc? I guess gcc-7 is enough. And I'm using gcc 10.

I will also look into solutions to avoid using c++17 to support old compilers.

**Comment by 71walceli** [comment]
Well, I compiled a program with the option:

```
apk@HP-ProBook-6470b ~/Projects/Data Structures and Algorithms $ g++ -std=c++17 -o 1.7.TemplateClass.cpp.out 1.7.TemplateClass.cpp 
apk@HP-ProBook-6470b ~/Projects/Data Structures and Algorithms $ ./1.7.TemplateClass.cpp.out Rectangle: width=10, height=20
Rectangle: width=20, height=10
Area: 200
Rectangle: width=11.5, height=22.5
Rectangle: width=22.5, height=11.5
Area: 258.75
```

What do you know! It worked with the option `-std=c++17`...

But cpprepl wont wotk... GCC 5 might not be fully compatible with it, might it?

**Comment by 71walceli** [comment]
> Is is possible for you to upgrade gcc? I guess gcc-7 is enough. And I'm using gcc 10.

I don't know, I have to fetch it elsewhere:

```
apk@HP-ProBook-6470b ~ $ sudo apt update 
...
apk@HP-ProBook-6470b ~ $ sudo apt install g++
Reading package lists... Done
Building dependency tree       
Reading state information... Done
g++ is already the newest version (4:5.3.1-1ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
apk@HP-ProBook-6470b ~ $ 

```

**Comment by Enter-tainer** [comment]
I found this stackoverflow page https://unix.stackexchange.com/a/272421 , hope that would be useful for you to install g++-10.

In short, you need to install g++-10, and then set it the default c++ compiler.

I'm not actully a Linux Mint user, so remember to recheck these shell commands before really running them, in case they break your system.

**Comment by 71walceli** [comment]
I'll try to have `cpprepl` it run...

Bad nows, `crepl` used to work fine,ultil yesterday that I updated.

```
l
✔ >>= · int *p=malloc(siseof(int)*5)
gcc: error: undefined: No such file or directory
✔ >>= · #include <stdlib.h>
gcc: error: undefined: No such file or directory
✔ >>= · int a
gcc: error: undefined: No such file or directory
✖ >>= · 
```

> I found this stackoverflow page https://unix.stackexchange.com/a/272421 , hope that would be useful for you to install g++-10.
> 
> In short, you need to install g++-10, and then set it the default c++ compiler.

I'll try updating gcc and g++. I might try setting up clang also

And thanks a lot! You'be been bery helpful so far!

**Comment by Enter-tainer** [comment]
crepl is indeed broken, I believe it is version 1.2.4 update that breaks it. I will later publish a fix.

**Comment by Enter-tainer** [comment]
ok, it's now fixed. pls install version 1.2.5(the latest).

**Comment by Enter-tainer** [comment]
Remember that this is a simple hobby project, and is kind of something like a proof of concept. It is not actively maintained. If you need a so called "real" repl, you should take a look at [cling](https://github.com/root-project/cling).

**Comment by 71walceli** [comment]
`crepl` works OK now.

`cpprepl` crashes wit segfault on my machine.

```
apk@HP-ProBook-6470b ~ $ cpprepl 
✔ >>= · int a=1
In file included from /tmp/repl.cxx:1:0:
/tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
 struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
                  ^
/tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
/tmp/dbg.h:142:63: error: template argument 2 is invalid
                                       std::declval<T>().end())>>
                                                               ^
Error: Command failed: /tmp/a.out
Segmentation fault (core dumped)

    at ChildProcess.exithandler (child_process.js:308:12)
    at ChildProcess.emit (events.js:314:20)
    at maybeClose (internal/child_process.js:1022:16)
    at Socket.<anonymous> (internal/child_process.js:444:11)
    at Socket.emit (events.js:314:20)
    at Pipe.<anonymous> (net.js:675:12) {
  killed: false,
  code: 139,
  signal: null,
  cmd: '/tmp/a.out',
  stdout: '',
  stderr: 'Segmentation fault (core dumped)\n'
}
```

> Remember that this is a simple hobby project, and is kind of something like a proof of concept. It is not actively maintained. If you need a so called "real" repl, you should take a look at [cling](https://github.com/root-project/cling).

Thanks for the `cling` advice. I understand however you decide about maintaining your project, Thanks a lot!

**Comment by 71walceli** [comment]
> `crepl` works OK now.
> 
> `cpprepl` crashes wit segfault on my machine.
> 
> ```
> apk@HP-ProBook-6470b ~ $ cpprepl 
> ✔ >>= · int a=1
> In file included from /tmp/repl.cxx:1:0:
> /tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
>  struct is_stl<T, std::void_t<decltype(std::declval<T>().begin(),
>                   ^
> /tmp/dbg.h:141:18: error: ‘void_t’ is not a member of ‘std’
> /tmp/dbg.h:142:63: error: template argument 2 is invalid
>                                        std::declval<T>().end())>>
>                                                                ^
> Error: Command failed: /tmp/a.out
> Segmentation fault (core dumped)
> 
>     at ChildProcess.exithandler (child_process.js:308:12)
>     at ChildProcess.emit (events.js:314:20)
>     at maybeClose (internal/child_process.js:1022:16)
>     at Socket.<anonymous> (internal/child_process.js:444:11)
>     at Socket.emit (events.js:314:20)
>     at Pipe.<anonymous> (net.js:675:12) {
>   killed: false,
>   code: 139,
>   signal: null,
>   cmd: '/tmp/a.out',
>   stdout: '',
>   stderr: 'Segmentation fault (core dumped)\n'
> }
> ```
> 
> > Remember that this is a simple hobby project, and is kind of something like a proof of concept. It is not actively maintained. If you need a so called "real" repl, you should take a look at [cling](https://github.com/root-project/cling).
> 
> Thanks for the `cling` advice. I understand however you decide about maintaining your project, Thanks a lot!

**Comment by 71walceli** [comment]
By the way, I didn't intend to close the issue!

**Comment by Enter-tainer** [comment]
reopened, can you run `g++ --version` on your machine?

**Comment by 71walceli** [comment]
> reopened, can you run `g++ --version` on your machine?

Sure

```
g++ (Ubuntu 5.5.0-12ubuntu1~16.04) 5.5.0 20171010
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by Enter-tainer** [comment]
OK, I will publish a version which support c++11, which supports your machine.

**Comment by 71walceli** [comment]
> I found this stackoverflow page https://unix.stackexchange.com/a/272421 , hope that would be useful for you to install g++-10.
> 
> In short, you need to install g++-10, and then set it the default c++ compiler.
> 
> I'm not actully a Linux Mint user, so remember to recheck these shell commands before really running them, in case they break your system.

**Comment by 71walceli** [comment]
> By the way, I didn't intend to close the issue! I can't reopen it...

Not again...! Excuse me...

**Comment by Enter-tainer** [comment]
1.2.6 is published, pls tell me if it is working on your machine

**Comment by 71walceli** [comment]
Now `crepl` and `cpprepl` work OK...

Have a look:

```
? >>= ‣ apk@HP-ProBook-6470b ~ $ cpprepl
✔ >>= · int a=1
✔ >>= · a++
a++ = 1 :: int
✔ >>= · a++
a++ = 2 :: int
✖ >>= · 
```

**Comment by 71walceli** [comment]
> Well, I compiled a program with the option:
> 
> ```
> apk@HP-ProBook-6470b ~/Projects/Data Structures and Algorithms $ g++ -std=c++17 -o 1.7.TemplateClass.cpp.out 1.7.TemplateClass.cpp 
> apk@HP-ProBook-6470b ~/Projects/Data Structures and Algorithms $ ./1.7.TemplateClass.cpp.out Rectangle: width=10, height=20
> Rectangle: width=20, height=10
> Area: 200
> Rectangle: width=11.5, height=22.5
> Rectangle: width=22.5, height=11.5
> Area: 258.75
> ```
> 
> What do you know! It worked with the option `-std=c++17`...

c++17 may not be fully implemented in g++ 5.5.0, but still, I wonder how this one compiled...

Source for 1.7.TemplateClass.cpp:
```cpp
#include <iostream>
#include <string>
#include <sstream>

template <class T>
class Rectangle
{
private:
  T width, height;
public:
  Rectangle(T width, T height);
  std::string toString();
  T area();
  void swapSides();
};

template <class T>
Rectangle<T>::Rectangle(T width, T height) {
  this->width  = width;
  this->height = height;
}

template <class T>
std::string Rectangle<T>::toString() {
  std::ostringstream o;
  o << "Rectangle: width="<< this->width <<", height="<< this->height;
  return o.str();
}
template <class T>
T Rectangle<T>::area() {
  return this->width*this->height;
}
template <class T>
void Rectangle<T>::swapSides() {
  T temp = this->height;
  this->height= this->width;
  this->width = temp;
}

int main(int argc, char const *argv[])
{
  Rectangle<int> r1(10,20);
  std::cout<< r1.toString()<<std::endl;
  r1.swapSides();
  std::cout<< r1.toString()<<std::endl;
  std::cout<<"Area: "<<r1.area()<<std::endl;
  
  Rectangle<float> r2(11.5,22.5);
  std::cout<< r2.toString()<<std::endl;
  r2.swapSides();
  std::cout<< r2.toString()<<std::endl;
  std::cout<<"Area: "<<r2.area()<<std::endl;
  
  return 0;
}
```

I'm just starting, so this code can't contain some advanced c++17 feature, I think,,,

**Comment by 71walceli** [comment]
> Now `crepl` and `cpprepl` work OK...
> 
> Have a look:
> 
> ```
> ? >>= ‣ apk@HP-ProBook-6470b ~ $ cpprepl
> ✔ >>= · int a=1
> ✔ >>= · a++
> a++ = 1 :: int
> ✔ >>= · a++
> a++ = 2 :: int
> ✖ >>= · 
> ```

So far, cpprepl works now...

**Comment by Enter-tainer** [comment]
> c++17 may not be fully implemented in g++ 5.5.0, but still, I wonder how this one compiled...

In short, my code use `std::void_t`(https://en.cppreference.com/w/cpp/types/void_t), and the error message said that "‘void_t’ is not a member of ‘std’", so I think maybe `std::void_t` is not in gcc 5's standard libaray.

gcc 5 only supports a few number of c++17 feature, you can checkout gcc's standard support page for details.

**Comment by 71walceli** [comment]
> > c++17 may not be fully implemented in g++ 5.5.0, but still, I wonder how this one compiled...
> 
> In short, my code use `std::void_t`(https://en.cppreference.com/w/cpp/types/void_t), and the error message said that "‘void_t’ is not a member of ‘std’", so I think maybe `std::void_t` is not in gcc 5's standard library.
> 
> gcc 5 only supports a few number of c++17 feature, you can checkout gcc's standard support page for details.

It's as I thought. The example I compiled didn't use those features simply `g++ ... -o ...` compiles it...

**Comment by 71walceli** [comment]
Thanks again...!
